### PR TITLE
Fix error reporting

### LINKED ISSUE #1220: Not found specialization does not pop up as a snackbar message
### Environment

- Scope: desktop IDE, browser IDE, CLI
- Version: master, latest, 079ab358e6d25eef913bd4f00ae2df7613c36ec6

### Steps to reproduce

1. File → Add Library → gweimer/utils
2. Place a `if-else-v` node
3. Bind `42` to `F`
4. Deploy → Show code for Arduino

### Expected behavior

- A clear snackbar message about missing specialization for `number` is shown
- Or at least a fallback message with the title “Unhandled error”, text in notes, and a solution like “The error was not caught, which is a bug. Report the issue to XOD developers.”

### Actual behavior

Nothing

### Console output

```
Uncaught Error: Can't find the patch in the project with specified path: "gweimer/utils/if-else-v(number)"
    at monads.js:102
    at index.js:2491
    at index.js:864
    at project.js:320
    at index.js:2491
    at index.js:860
    at _pipe.js:3
    at _pipe.js:3
    at _pipe.js:3
    at _arity.js:5
(anonymous) @ monads.js:102
(anonymous) @ index.js:2491
(anonymous) @ index.js:864
(anonymous) @ project.js:320
(anonymous) @ index.js:2491
(anonymous) @ index.js:860
(anonymous) @ _pipe.js:3
(anonymous) @ _pipe.js:3
(anonymous) @ _pipe.js:3
(anonymous) @ _arity.js:5
(anonymous) @ map.js:53
module.exports.XWrap.@@transducer/step @ _xwrap.js:10
_arrayReduce @ _reduce.js:11
_reduce @ _reduce.js:44
map @ map.js:52
(anonymous) @ _dispatchable.js:39
(anonymous) @ _curry2.js:20
f1 @ _curry1.js:17
(anonymous) @ _pipe.js:3
(anonymous) @ _arity.js:5
(anonymous) @ extractBoundInputsToConstNodes.js:41
(anonymous) @ index.js:2491
(anonymous) @ index.js:864
(anonymous) @ extractBoundInputsToConstNodes.js:293
(anonymous) @ index.js:2491
(anonymous) @ index.js:860
(anonymous) @ _pipe.js:3
(anonymous) @ _arity.js:5
_Right.chain @ Either.js:67
(anonymous) @ _dispatchable.js:30
(anonymous) @ _curry2.js:20
f1 @ _curry1.js:17
(anonymous) @ _pipe.js:3
(anonymous) @ _pipe.js:3
(anonymous) @ _pipe.js:3
(anonymous) @ _arity.js:5
(anonymous) @ transpiler.js:330
(anonymous) @ index.js:2491
(anonymous) @ index.js:868
(anonymous) @ transpiler.js:390
(anonymous) @ index.js:2491
(anonymous) @ index.js:864
(anonymous) @ App.jsx:60
(anonymous) @ Maybe.js:50
Just.reduce @ Maybe.js:141
(anonymous) @ Maybe.js:49
(anonymous) @ _curryN.js:37
(anonymous) @ _arity.js:18
transformProjectForTranspiler @ App.jsx:58
(anonymous) @ _pipe.js:3
(anonymous) @ _pipe.js:3
(anonymous) @ _arity.js:4
onShowCodeArduino @ App.jsx:43
apply @ /home/nailxx/devel/xod/node_modules/electron/dist/resources/electron.asar/common/api/callbacks-regi…:48
(anonymous) @ /home/nailxx/devel/xod/node_modules/electron/dist/resources/electron.asar/renderer/api/remote.js:299
emitThree @ events.js:116
emit @ events.js:194
```

### PR DESCRIPTION
**Pull request by evgenykochetkov**
- Closes #1220 
- Adds middleware to give user a chance to report unexpected errors

### CODE REVIEW

File: packages/xod-client/src/core/crashReporterMiddleware.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    return next(action);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+
+    return next(
+      addError(
+        composeMessage(
+          'You have found a bug, report it to XOD developers',
+          err.message
```
Comment (nkrkv) [inline packages/xod-client/src/core/crashReporterMiddleware.js]: Too long for a title. Maybe let it be more like the error without a formatter?

```js
return {
  title: 'Error',
  note: err.message,
  solution: 'The error was not caught, which is a bug. Report the issue to XOD developers.'
};
```

File: packages/xod-project/src/messages.js [hunk 2]
```diff
@@ -64,6 +64,78 @@ export default {
...
     )}`,
   }),
+  SPECIALIZATION_PATCH_CANT_BE_ABSTRACT: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: "Specialization patches can't be abstract",
+    trace,
+  }),
+  SPECIALIZATION_PATCH_MUST_HAVE_SAME_ARITY_LEVEL: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: 'Specialization patches must have same arity level as abstract',
```
Comment (nkrkv) [inline packages/xod-project/src/messages.js]: Perhaps you mean “arity step,” not “arity level?”

Nevertheless, more details are necessary:

> Specialization patches must have same arity step as abstract. However, `@/foo(number)` has step 1, whereas `@/foo` has 2.

solution:

> Change `variadic-1` marker to `variadic-2` on specialization to continue.

File: packages/xod-project/src/messages.js [hunk 3]
```diff
@@ -64,6 +64,78 @@ export default {
       conflictingSpecializations
     )}`,
   }),
+  SPECIALIZATION_PATCH_CANT_BE_ABSTRACT: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: "Specialization patches can't be abstract",
+    trace,
```
Comment (nkrkv) [inline packages/xod-project/src/messages.js]: Solution:

> Remove the `abstract` marker node from `@/foo(number)` and provide a concrete implementation in XOD or C++.

File: packages/xod-project/src/messages.js [hunk 4]
```diff
@@ -64,6 +64,78 @@ export default {
...
+  }),
+  SPECIALIZATION_PATCH_MUST_HAVE_SAME_ARITY_LEVEL: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: 'Specialization patches must have same arity level as abstract',
+    trace,
+  }),
+  SPECIALIZATION_PATCH_CANT_HAVE_GENERIC_PINS: ({ trace }) => ({
+    title: 'Invalid specialization patch',
+    note: "Specialization patches can't have generic pins",
+    trace,
```
Comment (nkrkv) [inline packages/xod-project/src/messages.js]: Note.

> ... Found generic pins: `IN1`.

Solution:

> Replace generic terminal nodes in `@/foo(number)` with specific ones.

File: packages/xod-project/src/messages.js [hunk 5]
```diff
@@ -188,6 +260,10 @@ export default {
     note: 'The program has a cycle',
     solution: 'Use xod/core/defer node to break the cycle',
   }),
+  BAD_LITERAL_VALUE: ({ value }) => ({
+    title: 'Bad literal value',
+    note: `${value} is not a valid literal`,
```
Comment (nkrkv) [inline packages/xod-project/src/messages.js]: Solution is:

`If you meant a string, surround it with double quotes: "${literal}".`

Check, however `packages/xod-project/src/messages.js` → `INVALID_LITERAL_BOUND_TO_PIN`. Looks like code duplication.

File: packages/xod-project/src/messages.js [hunk 6]
```diff
@@ -64,6 +64,78 @@ export default {
...
+    title: 'Invalid specialization patch',
+    note: `Conflicting types for generic ${genericType}: ${typeNames}`,
+    trace,
+  }),
+  SPECIALIZATION_HAS_WRONG_NAME: ({
+    expectedSpecializationBaseName,
+    trace,
+  }) => ({
+    title: 'Invalid specialization patch',
+    note: `Patch should be named ${expectedSpecializationBaseName}`,
```
Comment (nkrkv) [inline packages/xod-project/src/messages.js]: Here and above, solutions are missing. We should provide them to be a friendlier environment.

File: packages/xod-project/src/messages.js (hunk 5 above)
Comment (evgenykochetkov) [reply packages/xod-project/src/messages.js]: `BAD_LITERAL_VALUE` is returned from `getTypeFromLiteral` which takes just a `DataValue` and does not know about pins etc.

**Comment by evgenykochetkov** [comment]
Specialization validation errors will be dealt with in a separate issue: #1231

File: packages/xod-client/src/core/crashReporterMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    return next(action);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+
+    return next(
+      addError(
+        composeMessage(
+          'Error',
+          err.message,
```
Comment (nkrkv) [inline packages/xod-client/src/core/crashReporterMiddleware.js]: Oops. The `composeMessage` is legacy. Better directly construct the stanza. It’s more readable:

```js
{
  title: 'Error',
  note: err.message,
  solution: 'The error was not caught, which is a bug. Report the issue to XOD developers.'
}
```

**Pull request merged**